Parking Instant Answer

AlexCutlipParkingPanda opened the pull request:
**What does your Instant Answer do?**
This IA queries the Parking Panda location search api for paid parking near locations in the search string. 

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Similar to the Places IA when searching for hotels this Instant Answer provides an overview of available parking options for a location as well as a way to reserve parking in advance.

**What is the data source for your Instant Answer? (Provide a link if possible)**
The Parking Panda API https://www.parkingpanda.com/api/v2/help

**Why did you choose this data source?**
I chose this data source because it is ease of use, I am familiar with it, I know the data to be high quality and because I work for Parking Panda.

**Are there any other alternative (better) data sources?**
There are other parking reservation services with APIs such as ParkWhiz and ParkMe.

**What are some example queries that trigger this Instant Answer?**
"where to park near fenway park", "parking near radio city music hall"

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Anyone planing to drive to somewhere with limited parking.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No.

**Which existing Instant Answers will this one supersede/overlap with?**
There is overlap with the Places IA for locations where there are parking reviews on yelp. However, I think this IA provides more useful results.
Here is an example https://duckduckgo.com/?q=parking+garages+in+new+york+city&ia=places.

**Are you having any problems? Do you need our help with anything?**
There are searches which include the word parking (or another trigger word) but not a location (e.g "domain parking" or "parking tickets") for which this spice triggers. Currently the API is queried and no results are returned but I wonder if there is a better way to handle this.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
The Parking Panda API is public and free to use.

**Where did you hear about DuckDuckHack? (For first time contributors)**
Zac Pappis at DuckDuckGo.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
![list](https://cloud.githubusercontent.com/assets/10423789/5803849/65fbf7e2-9fcf-11e4-839d-f74ac7df80e9.PNG)
![single_item](https://cloud.githubusercontent.com/assets/10423789/5803848/65fb3e2e-9fcf-11e4-9392-303926c7e96c.PNG)


## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [X] Google Chrome
        [X] Firefox
        [X] Opera
        [X] IE 10

    - Windows 7
        [X] Google Chrome
        [X] Firefox
        [X] Opera
        [-] IE 8 (There is an issue with IE8 and the 'products_item_detail' template)
        [X] IE 9
        [X] IE 10

    - Windows XP
        [Very Broken] IE 7
        [Same issue as Win7] IE 8 

    - Mac OSX
        [X] Google Chrome
        [X] Firefox
        [X] Opera
        [X] Safari

    - iOS (iPhone)
        [X] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [X] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [X] Firefox
        [] Native Browser
        [X] Google Chrome
        [] Opera
```

jagtalon commented:
@AlexCutlipParkingPanda nice, we'll check it out. About the button though, we only go with these colors:

![screen shot 2015-01-19 at 5 23 01 pm](https://cloud.githubusercontent.com/assets/81969/5808841/eef744d6-9fff-11e4-8cb8-903024bc6e34.png)
![screen shot 2015-01-19 at 5 23 23 pm](https://cloud.githubusercontent.com/assets/81969/5808842/eef84868-9fff-11e4-8e51-7d3923720aad.png)

jagtalon commented on share/spice/parking/parking.js:
```diff
@@ -0,0 +1,111 @@
+    env.ddg_spice_parking = function(api_result){
+        
+        // Check that results were returned successfully
+        if (!api_result.success || api_result.resultsCount === 0) {
```
Have to check if `api_result` exists before accessing anything, i.e., `if(!api_result || !api_result.success || ...`

AlexCutlipParkingPanda commented:
Hi @jagtalon!  Thanks for the quick review! 
I just pushed a commit changing the button color and fixing that line of JS.

jagtalon commented:
Thanks @AlexCutlipParkingPanda checking out the code! 

In the mean time, what are the terms or the API? Is it okay to use commercially? Will we have to pay?

jagtalon commented:
@AlexCutlipParkingPanda I think a map would be better for this one:

![screen shot 2015-01-20 at 11 06 59 am](https://cloud.githubusercontent.com/assets/81969/5820665/80f09dd0-a094-11e4-8b01-9f47a04ae755.png)

What do you think, @abeyang?

abeyang commented:
I agree with @jagtalon

AlexCutlipParkingPanda commented:
@jagtalon There is currently no separate terms of use for the Parking Panda API. Use would just be governed generally by the ParkingPanda.com terms of use. The API is free and okay to use commercially.

AlexCutlipParkingPanda commented:
@jagtalon @abeyang I also agree that this data would be better communicated via a map.

When I was implementing this initially I didn't realize it was possible to create a spice that emulated the Places IA. I was working off of https://duck.co/duckduckhack/spice_templates_overview rather than https://duck.co/duckduckhack/spice_displaying.

I also attempted to convert to the places template but found it rather rigid in the following ways.
a) It seems that price is only displayed as 0-4 dollar signs. I would prefer to set a price string.
b) I had to hijack the neighborhood property to display the distance, the reviews property to display the price on the front of the tile and the address property to display the description. I would like to specify sub templates so that future updates to the Places template won't cause the Parking IA to change in unexpected ways.
c) I would like to make the sourceName text in the tile different than the sourceName text used in the metaBar. 

Here's a screenshot of where I left off.
![image](https://cloud.githubusercontent.com/assets/10423789/5825382/e2007914-a0b8-11e4-97ad-509ad82d0baa.png)

If you all think it's a good idea, I would like to work with you to further generalize the Places  template and to convert this IA over to it.

In the meantime, would it make more sense to move forward with the current Parking IA then later submit a Places version as an improvement or would it be better to switch to using Places now?

moollaza commented:
@AlexCutlipParkingPanda hey there, we've been talking internally and it looks like you should be able to pass along a custom Tile template to the Places view, so you won't have to force other info into the default Places tile.

We're playing around with it internally to see how it will work and how we'd like it to look. We'll get back to your shortly and we can go from there.

Thanks again!

bsstoner commented on share/spice/parking/parking.js:
```diff
@@ -0,0 +1,111 @@
+        }
+        
+        // Render the response
+        Spice.add({
```
Unfortunately right now you need to require the 'maps' module before the call to Spice.add(). I'd like to take care of this transparently eventually (like we do with audio), but for now you can just do this:
```
DDG.require('maps', function() {
  Spice.add(...)
});
```

moollaza commented:
@AlexCutlipParkingPanda sorry for the delay, we'll have a design for you very soon! Can't wait to see this go live :smiley:

AlexCutlipParkingPanda commented:
@moollaza I look forward to seeing what you all come up with!

AlexCutlipParkingPanda commented:
@moollaza Just checking in to see if there have been any updates on this front.

moollaza commented:
@AlexCutlipParkingPanda apologies for the delay, we do have some awesome updates for you!

Our design team has created a couple mockups that demonstrate the end-goal for this IA:

**Tile Front:**
![parking-v1 1-a-initial](https://cloud.githubusercontent.com/assets/873785/6010509/0b8ed076-ab01-11e4-8c87-a6d2d59404e2.jpg)

**Tile Back:**
![parking-v1 1-b-tile-active](https://cloud.githubusercontent.com/assets/873785/6010508/09669cf2-ab01-11e4-90a1-eb89a75510f1.jpg)

------

You may have noticed the mockup includes a static marker showing the location of the venue that the user is trying to park near. Right now, our engineering team is working on a way to implement that feature so that it's possible (we haven't done an integration like this before) but for now if you could implement the tile view in the mockup it would be great.

Seeing as this is a special custom design, it will require a custom tile template. Your code to render the Spice should look like this:

```javascript
DDG.require('maps', function() {
    Spice.add({
        id: "parking",
        name: "Parking",
        model: 'Place',
        view: 'Places',
        data: api_result.data.locations,
        meta: {
            primaryText: "Parking Near: " + api_result.data.search.displayText,                
            sourceName: "ParkingPanda.com",
            sourceUrl: 'https://www.parkingpanda.com/Search/?ref=duckduck&location=' + api_result.data.search.query
        },
        normalize: normalize,
        templates: {                
            group: 'places',
            item: Spice.parking.item
        }
    });
});
```

And the template code (which will go into `/share/spice/parking/item.handlebars`), will need to look like this:

```html
<div class="tile  tile--f  tile--loc">

    <!-- Front of tiel -->
    <div class="tile--f__main  tile--loc__main">
        <div class="tile__media">
            {{#if image}}
                <div class="tile__media__wrapper">
                    <img src="" data-src="{{imageProxy image}}" alt="{{title}}" class="tile__media__img  js-lazyload" />
                </div>
            {{else}}
                <span class="ddgsi tile__media__no-img">@</span>
            {{/if}}
        </div>
        <div class="tile__num">{{num}}</div>

        <!-- Update tile_body, tile_title and tile_foot as needed -->
        <div class="tile__body">
            <h6 class="tile__title">{{ellipsis displayName 33}}</h6>
            <div class="tile__foot"></div>
        </div>
    </div>

    <!-- Back of tile -->
    <div class="tile--f__alt  tile--loc__alt">
        <div class="tile__body  {{#and url meta.sourceName}}has-foot{{/and}}">

            <!-- Update as needed -->
            <h6 class="tile__title">
                <a href="{{url}}" title="{{name}}">{{ellipsis displayName 33}}</a>
            </h6>

            {{#if price}}
                <span class="tile__price">{{{priceSymbols price 4}}}</span>
                <span class="tile__price">{{price}}+</span>
            {{/if}}

            {{#if address_lines}}
                <a class="tile__address" href="{{directions}}">
                    {{#each address_lines}}
                        {{.}}<br />
                    {{/each}}
                </a>
            {{else}}
                {{#if address}}
                    <span class="tile__address">{{ellipsis address 65}}</span>
                {{/if}}
            {{/if}}

            {{#if phone}}
                <a class="tile__phone" href="tel:{{phone}}">{{phone}}</a>
            {{/if}}

            {{#and url meta.sourceName}}
                <div class="tile__foot">
                    {{{moreAt url meta.sourceName dynamicMoreAtText=5}}}
                </div>
            {{/and}}
        </div>
    </div>
</div>
```

Again, I'm very sorry about the delay, I can't wait to get this live though!
Please let me know if you guys have any questions or need any help.

Thanks!

abeyang commented:
@AlexCutlipParkingPanda actually it's going to look more like this (red pin and blue button in the back):

![parking-v1 2](https://cloud.githubusercontent.com/assets/794084/6012887/8fff7756-ab02-11e4-9888-4fe068c2c52e.jpg)

Like @moollaza, I can't wait to see this live as well!

moollaza commented:
@AlexCutlipParkingPanda please ping us once the tiles are implemented! Don't worry about the extra pin for now, we'll add that in later for you, once we are able to :)

Feel free to ping me if you have any questions or need help with anything.

AlexCutlipParkingPanda commented:
Hi @moollaza !
The new template based places looks great! I've finished an initial rough conversion of the parking IA.

![image](https://cloud.githubusercontent.com/assets/10423789/6050030/5fca354c-ac8a-11e4-90bd-108ff5e0d87a.png)

I thought it would be best to wait for the search location pin before doing another round of thorough cross platform testing but in my cursory testing I noticed that click events on anchors within the tiles are bubbling up to the tile div on all browsers other than Chrome. Is there a way to get to request non-minified JS to make make debugging this a bit easier?

I also have a question about how to best handle optional and dynamic parameters in the 'spice to' string. I would like to not include defaultSearch if the user's local time cannot be determined. I also wonder how best to test this. Is it possible to pass regex to test_spice()? Would it be best to simply avoid sending the time?

Please let me know what you think and thanks again for all the help!

moollaza commented:
Hey @AlexCutlipParkingPanda, thanks a lot for implementing the mockup this is looking excellent so far!

>I noticed that click events on anchors within the tiles are bubbling up to the tile div on all browsers other than Chrome.

Yes, I'm able to reproduce this. Not sure why it's happening but I'll investigate it further.

> Is there a way to get to request non-minified JS to make make debugging this a bit easier?

Not yet, we're hoping to bring unminified JS and better logging to developers shortly. Sorry!

> I would like to not include defaultSearch if the user's local time cannot be determined.

It looks like your code is already doing this? Do you mean you want to not pass along that parameter at at all?

Unfortunately we don't currently have a great way to handle dynamic parameters in the URL. Can I ask what the `defaultSearch` param is for? It seems like removing it doesn't affect the results?

>I also wonder how best to test this. Is it possible to pass regex to test_spice()? Would it be best to simply avoid sending the time?

I actually have an open pull request to add this functionality for Spice but it still needs some testing and review.  Writing Spice tests that depend on location data is covered in our docs [here](https://duck.co/duckduckhack/testing_location_api). You can indicate what the location is for the test and write the test so that it can always expect the same result. Likewise you can define your own location that doesn't have a `time_zone` and write a test to make sure the resulting API call is correct.

If we can do without passing along the time for now, it would certainly make things a little simpler and then we can keep moving forward with this Instant Answer :)

------

Thanks again for doing this. Everything seems pretty good so far, so I'll run this past our team and I'll be sure to get back to you about the click handlers.

Cheers

AlexCutlipParkingPanda commented:
Hi @moollaza!

>>I would like to not include defaultSearch if the user's local time cannot be determined.

>It looks like your code is already doing this? Do you mean you want to not pass along that parameter at at all?

Yeah, I would like to not include it at all if I don't have anything to pass in. Something like a parameter hash could be neat here. 

>Can I ask what the `defaultSearch` param is for? It seems like removing it doesn't affect the results?

My goal with the `defaultSearch` parameter is to have api searches which don't specify specific times return the same results as the website. Currently an api search lacking times executes an 'all-day' search for the current day while the website returns results for the next three hours. This features is not yet in production so you won't see any change in results yet.

>If we can do without passing along the time for now, it would certainly make things a little simpler and then we can keep moving forward with this Instant Answer :)

I've changed `defaultSearch` to take times zones and updated the IA and tests accordingly. 

>Thanks again for doing this. Everything seems pretty good so far, so I'll run this past our team and I'll be sure to get back to you about the click handlers.

Thank you for going out of your way to help with Places and make this IA awesome! Please let me know when click handling or the search location marker is ready!

moollaza commented:
Hi @AlexCutlipParkingPanda, thanks for the updates and explanations.

This is looking great! Regarding the click events, we've deployed the IA to a testing machine and everything works correctly so there must be a small bug with the DuckPAN JS (I'll be sure to investigate!):

https://ddh5.duckduckgo.com/?q=parking+san+francisco&ia=parking&iax=1

The only other thing I noticed that should be fixed is the `sourceName`; Would you mind removing the `.com` from "ParkingPanda.com"?

@abeyang how's it looking? (we haven't implemented the location pin yet, but that's an internal change)

abeyang commented:
It's looking great, @AlexCutlipParkingPanda -- almost there! But don't we want to use the **garage name** by default, as opposed to the street address? (Or is it the case that these lots don't have names?)

![parking](https://cloud.githubusercontent.com/assets/794084/6118141/68c8fde8-b06f-11e4-86cb-aadaf13d7234.png)

Moreover, can you make these css tweaks to the "reserve parking" button:

``` css
#zci-parking .reserve-btn--parking {
/* remove bottom border */
  border-bottom: 0;

/* button should fill up inside the tile */
  width: 100%;
  padding: .1em 0;
}
```

The resulting button should look like this:
![screen shot 2015-02-09 at 3 24 54 pm](https://cloud.githubusercontent.com/assets/794084/6118173/d3855d2a-b06f-11e4-9068-3c954096ed9a.png)

AlexCutlipParkingPanda commented:
@abeyang Thanks for the feedback! I just pushed the changes you recommended.

With regards to the garage names your intuition is correct. Many locations do not have public names. 
Here is an example of a few named locaitons: https://ddh5.duckduckgo.com/?q=parking+baltimore+md

jagtalon commented:
@AlexCutlipParkingPanda Please change the protocol in `spice to` from HTTP to HTTPS. The endpoint redirects, and our backend doesn't allow any redirection. Changing to HTTPS should prevent any redirects. :)

AlexCutlipParkingPanda commented:
@jagtalon Thanks for letting me know! Updated.

AlexCutlipParkingPanda commented:
@jagtalon Do you know what remains to be done to move forward with this? Is the search location pin the last thing left?
Please let me know.
Thanks!

moollaza commented:
@AlexCutlipParkingPanda sorry about the delay, this LGTM :+1: 

We'll add the search location pin later :smiley:

moollaza commented:
Updated on my testing machine: https://ddh7.duckduckgo.com/?q=parking+near+radio+city+music+hall&ia=parking&iax=1

moollaza commented:
@AlexCutlipParkingPanda just wanted to say thanks again for the awesome job. We'll slate this for release and it should go live next week. We'll be sure to notify you when it goes live

AlexCutlipParkingPanda commented:
@moollaza @jagtalon @abeyang 
Thanks for all the support!
It's been a pleasure working with you all!

moollaza commented:
@AlexCutlipParkingPanda apologies for the delay, this one will be live this week, I'll let you know as soon as it is!

> It's been a pleasure working with you all!

It was great working with you as well!

moollaza commented:
@AlexCutlipParkingPanda apologies for the delay, but it's finally live!

https://duckduckgo.com/?q=parking+near+radio+city+music+hall&ia=parking

Thanks again :smiley:

AlexCutlipParkingPanda commented:
@moollaza Awesome! :smile:
Thanks again for all the help!

With regards to the search location pin: Will you drop me a note when it's ready for implementation or is there somewhere I can check on its status?

moollaza commented:
@AlexCutlipParkingPanda I'll ping you once it's ready for implementation. We have some internal tasks to create the functionality.

AlexCutlipParkingPanda commented:
@moollaza It looks like something's wrong with the tiles. "is-selected" is not getting added to ".tile--f" on click and the tiles therefore never 'flip'.
This looks like it's affecting all places AIs: 
https://duckduckgo.com/?q=pizza+near+me&ia=places&iax=1

arthurtalpaert commented:
Hello! I didn't find a more appropriate place to report a bug. For me, the instant answer doesn't work and looks broken, see screenshot:
![screenshot from 2015-04-07 16 43 04](https://cloud.githubusercontent.com/assets/5888770/7025751/f5ac25d4-dd45-11e4-9b74-baa6a1834a7d.png)

I am typing this from Paris, France, if this has any impact.
Cheers!

moollaza commented:
@arthurtalpaert typically bugs are reported by making a new issue and indicating the name of the related Instant Answer.

Can you please make a GitHub issue for this and also add a pic of your network requests, or let me know if you see a call to `/js/spice/parking/<foo>` -- I need to see what the API call is returning, if anything, to see what's causing the problem.

Thanks for taking the time to report this! We really appreciate it :+1:
